File: packages/snow_draw_core/lib/draw/history/recordable.dart
/// Marker interface for actions that should be recorded in history.
abstract interface class Recordable {
  String get historyDescription;
  HistoryRecordType get recordType;
}

/// Marker interface for actions that should not be recorded.
abstract interface class NonRecordable {
  String get nonRecordableReason;
}

/// Categories for recorded history entries.
enum HistoryRecordType { edit, create, delete, style, selection, other }

File: packages/snow_draw_rust_engine/src/draw/history/recordable.rs
#![allow(dead_code)]

/// Marker trait for actions that should be recorded in history.
pub trait Recordable {
    /// Human-readable description for this history entry.
    fn history_description(&self) -> String;

    /// Category used by history policies and UI grouping.
    fn record_type(&self) -> HistoryRecordType;
}

/// Categories for recorded history entries.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
pub enum HistoryRecordType {
    Edit,
    Create,
    Delete,
    Style,
    Selection,
    #[default]
    Other,
}
